### PR TITLE
CC-7247: allow topics.dir config to be more configurable

### PR DESCRIPTION
**Pull request by levzem**
see CC ticket and relevant one pager for explanation

### CODE REVIEW

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 1]
```diff
@@ -458,6 +550,22 @@ public static ConfigDef getConfig() {
         FORMAT_CLASS_RECOMMENDER
     );
 
+    visible.define(
+        "topics.dir",
+        Type.STRING,
+        "topics",
+        Importance.HIGH,
+        "Top level directory to store the data ingested from Kafka. "
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Make these constants?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 2]
```diff
@@ -369,6 +415,51 @@ public String getUrl() {
...
+      List<String> topicGroups = new ArrayList<>();
+      Matcher matcher = topixRegexCaptureGroup.matcher(topic);
+      while (matcher.find()) {
+        if (!matcher.group().isEmpty()) {
+          topicGroups.add(matcher.group());
+        }
+      }
+
+      // make sure that all references to captured groups actually exist
+      int maxVar = extractMaximumInt(topicsDir);
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: I don't understand this part? I thought topicsDir can contain numbers in them - "/topic_1000" by itself should be a valid topicDir? Or if the topic name contains numbers and ${topic} appears in topicDir, wouldn't your maxVar be not what you expect? I think you can do this with a regex looking for anything that looks like "\$\{\d\}" in your topicsDir and replace with what you find in topicGroups. Catch `IndexOutOfBoundsException` and rethrow the `ConfigException` for referencing non-existing capture group.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 3]
```diff
@@ -469,6 +577,32 @@ private static void addAllConfigKeys(ConfigDef container, ConfigDef other, Set<S
...
 
+  /**
+   * Extract the biggest integer contained in the string
+   *
+   * @param string - the string to extract the maximum integer from
+   * @return int - the biggest integer found in the string, 0 if none found
+   */
+  private static int extractMaximumInt(String string) {
+    int num = 0;
+    int res = 0;
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Integer overflow? I actually don't think you need this method at all.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 1 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: big derp :P

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 2 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: thats actually a lot smarter
thanks!

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 4]
```diff
@@ -92,7 +96,7 @@
   public static final String TOPIC_REGEX_CAPTURE_GROUP_DOC = "A regex that specifies what groups "
       + "to capture in the topic, so when specifying `${1}` in `topics.dir`, `${1}` will refer to "
       + "the first captured group. Example config value of `[a-zA-Z]*` will capture all characters "
-      + "delimited by . _  and -, so for `topic.dir = ${1}/${2}` and `topic = topic.name.suffix` "
+      + "between [-._0-9], so for `topic.dir = ${1}/${2}` and `topic = topic.name_suffix` "
       + "the corresponding `topic.dir` will be `topic/name/`. By default, this functionality is "
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Since the parameter is called `topic`, maybe it would be less confusing to not have the string "topic" appear again in the example? Something like "example_name.suffix" -> "example/name" ?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 5]
```diff
@@ -551,19 +560,18 @@ public static ConfigDef getConfig() {
...
         Type.STRING,
-        "topics",
+        TOPICS_DIR_DEFAULT,
         Importance.HIGH,
-        "Top level directory to store the data ingested from Kafka. "
-            + "Supports ``${topic}`` in the value, which will be replaced by the actual"
-            + " topic name. Supports ``${1}``, ..., ``${n}`` in conjunction with "
-            + TOPIC_REGEX_CAPTURE_GROUP_CONFIG + ". See " + TOPIC_REGEX_CAPTURE_GROUP_CONFIG
-            + " configuration documentation for details.",
+        TOPICS_DIR_DOC + " Supports ``${topic}`` in the value, which will be "
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Why this doc string is not a constant like the rest?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 6]
```diff
@@ -578,29 +586,26 @@ private static void addAllConfigKeys(ConfigDef container, ConfigDef other, Set<S
...
+   * @return list of all numbers found in the regex pattern
    */
-  private static int extractMaximumInt(String string) {
-    int num = 0;
-    int res = 0;
+  private List<Integer> getPartsToReplace() {
 
-    for (char c : string.toCharArray()) {
+    List<Integer> toReplace = new ArrayList<>();
+    Pattern pattern = Pattern.compile("\\$\\{\\d+}");
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Do you need to escape the end bracket? Looks weird with mismatched escapes :-/

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 7]
```diff
@@ -578,29 +586,26 @@ private static void addAllConfigKeys(ConfigDef container, ConfigDef other, Set<S
...
-        num = num * 10 + (c - '0');
-      } else {
-        res = Math.max(res, num);
-
-        // Reset the number
-        num = 0;
+    while (partsMatcher.find()) {
+      String part = partsMatcher.group();
+      if (!part.isEmpty()) {
+        part = part.substring(2, part.length() - 1); // trim "${" and "}"
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: You can use match groups for this. With something like "\\$\\{(\d+)\\}" as pattern, match group 1 will be the digits only.

File: src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfigTest.java [hunk 8]
```diff
@@ -84,6 +84,21 @@ public void testTopicDirFromTopicParts() {
     );
   }
 
+  @Test
+  public void testTopicDirCanContainNumber() {
```
Comment (ncliang) [inline src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfigTest.java]: Add a testcase to test something that is smaller than the range of valid match groups - `${0}`, `${-1000}` or something.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 9]
```diff
@@ -578,29 +586,26 @@ private static void addAllConfigKeys(ConfigDef container, ConfigDef other, Set<S
...
+        toReplace.add(Integer.valueOf(part));
       }
     }
 
-    return Math.max(res, num);
+    Collections.sort(toReplace);
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: No need to sort here just to get the max. Just use `Collections.max()` outside in the caller. Makes it clearer what the code outside is doing, is slightly faster, and doesn't rely on this method returning a sorted list.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 10]
```diff
@@ -360,6 +365,8 @@ protected HdfsSinkConnectorConfig(ConfigDef configDef, Map<String, String> props
           e
       );
     }
+
+    partsToReplace = getPartsToReplace();
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Why does `partsToReplace` need to be a field?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 11]
```diff
@@ -438,22 +445,24 @@ public String getTopicDirFromTopic(String topic) {
...
+                  topicGroups.size()
+              )
+          );
+        }
       }
 
-      for (int i = 0; i < maxVar; i++) {
-        topicsDir = topicsDir.replace("${" + (i + 1) + "}", topicGroups.get(i));
+      for (int index : partsToReplace) {
+        topicsDir = topicsDir.replace("${" + index + "}", topicGroups.get(index - 1));
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Another way to do this that might be even nicer is to make `topicGroups` a Map<String, String> mapping "${1}" -> "captured_value_1", "${2}" -> "captured_value_2", etc. Then you can do something like 
```
topicGroups.forEach((k, v) -> topicsDir.replaceAll(k, v));
```
If you still see something that looks like "${\d+}" after that, user tried to reference invalid capture group, throw.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 6 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: its redundant to escape the second bracket - i can but i get a warning

File: src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfigTest.java (hunk 8 above)
Comment (levzem) [reply src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfigTest.java]: great edge case!

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 10 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: micro optimization to not compile the regex pattern and perform the regex search every call, but I just got a better idea - Ill just keep track of the largest index instead

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 11 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: I changed the implementation + behavior slightly (see one pager for full explanation) so the code is slightly different not, but let me know what you think

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 5 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: fixed

File: src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfigTest.java [hunk 12]
```diff
@@ -53,6 +54,107 @@ public void setUp() throws Exception {
...
+    properties.put(StorageCommonConfig.TOPICS_DIR_CONFIG, topicDir);
+    connectorConfig = new HdfsSinkConnectorConfig(properties);
+
+    connectorConfig.getTopicsDirFromTopic(topic);
+  }
+
+  @Test(expected = ConfigException.class)
+  public void testInvalidTopicDirNegative() {
+    String topic = "a.b.c.d";
+    String topicDir = "${-1}-${2}-${3}-${4}";
```
Comment (ncliang) [inline src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfigTest.java]: Maybe another case with ${0}. I think that would be the more common mistake.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 13]
```diff
@@ -399,6 +463,143 @@ public Configuration getHadoopConfiguration() {
...
+   * @param config - String - the config to validate
+   */
+  private void validateReplacements(String config) {
+    // remove all valid ${} substitutions
+    Pattern pattern = Pattern.compile("\\$\\{(\\d+)}");
+    Matcher partsMatcher = pattern.matcher(getString(config));
+    String dir = partsMatcher.replaceAll("").replace("${topic}", "");
+
+    // check for invalid ${} substitutions
+    Pattern invalid = Pattern.compile("\\$\\{.*}");
```
Comment (ncliang) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: nit: these patterns can be made `private static final` and reused.

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 14]
```diff
@@ -75,12 +84,30 @@
...
+  public static final String TOPIC_REGEX_CAPTURE_GROUP_DOC = "A regex that matches the entire "
+      + "topic and  specifies what groups to capture in the topic, so when specifying `${1}` in "
+      + "`topics.dir`, `${1}` will refer to the first captured group. Example config value of "
+      + "`([a-zA-Z]*)_([a-zA-Z]*)` match a topic that is two words delimited by a - and will "
+      + "capture both words as separate groupds, so for `topic.dir = ${1}/${2}` and "
+      + "`topic = example_name` the corresponding `topic.dir` will be `example/name/`. By default, "
+      + "this functionality is not enabled.";
+  public static final String TOPIC_REGEX_CAPTURE_GROUP_DEFAULT = null;
+
+  private static final String DIR_REGEX_DOC = " Supports ``${topic}`` in the value, which will be "
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: `${0}` and `${topic}` would be the same if the regex matched the whole topic , which we could choose to enforce. I guess the cost of adding `${topic}` is pretty low, and this feature could be enabled without ever specifying a capture config like `^.*$`

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 15]
```diff
@@ -281,9 +308,28 @@ public static ConfigDef newConfigDef() {
         FORMAT_CLASS_RECOMMENDER,
         AVRO_COMPRESSION_RECOMMENDER);
 
+    int lastOrder = 0;
+    String group = "Storage";
     for (ConfigDef.ConfigKey key : storageConfigDef.configKeys().values()) {
       configDef.define(key);
+      group = key.group;
+      lastOrder = key.orderInGroup;
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Does `.values()` always iterate over these in-order?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 16]
```diff
@@ -314,6 +363,21 @@ protected HdfsSinkConnectorConfig(ConfigDef configDef, Map<String, String> props
     addToGlobal(commonConfig);
     addToGlobal(this);
     this.url = extractUrl();
+    try {
+      this.topixRegexCaptureGroup = getString(TOPIC_REGEX_CAPTURE_GROUP_CONFIG) != null
+          ? Pattern.compile(getString(TOPIC_REGEX_CAPTURE_GROUP_CONFIG))
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Pull this out to a variable

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 17]
```diff
@@ -399,6 +463,143 @@ public Configuration getHadoopConfiguration() {
...
   }
 
+  /**
+   * Performs all substitutions on `logs.dir` and calculates the final directory for a topic
+   *
+   * @param topic - String - the topic whose directory to find
+   * @return String - the directory name and path
+   */
+  public String getLogsDirFromTopic(String topic) {
+    return getDirFromTopic(getString(LOGS_DIR_CONFIG), topic, logDirGroupsMaxIndex);
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Have you verified that you've replaced all of the call-sites for `getString(LOGS_DIR_CONFIG)` and `getString(TOPICS_DIR_CONFIG)` with your new function?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 18]
```diff
@@ -399,6 +463,143 @@ public Configuration getHadoopConfiguration() {
...
+  private String getDirFromTopic(String dir, String topic, int maxIndexParts) {
+    dir = dir.replace("${topic}", topic);
+
+    // only if configured
+    if (topixRegexCaptureGroup != null) {
+
+      // find all of the captured groups by the regex
+      Matcher matcher = topixRegexCaptureGroup.matcher(topic);
+      if (!matcher.matches()) {
+        throw new ConfigException(
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: You can provide the configuration and its value to the `ConfigException` for better formatting of the error.
```suggestion
        throw new ConfigException(
            TOPIC_REGEX_CAPTURE_GROUP_CONFIG,
            topixRegexCaptureGroup.pattern(),
```

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 19]
```diff
@@ -399,6 +463,143 @@ public Configuration getHadoopConfiguration() {
...
+                    + "but actually had %d capture groups.",
+                topic,
+                maxIndexParts,
+                topixRegexCaptureGroup.pattern(),
+                matcher.groupCount()
+            )
+        );
+      }
+
+      for (int index = 1; index < matcher.groupCount() + 1; index++) {
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Might as well give them all the groups
```suggestion
      for (int index = 0; index < matcher.groupCount() + 1; index++) {
```

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 20]
```diff
@@ -399,6 +463,143 @@ public Configuration getHadoopConfiguration() {
...
+    Matcher partsMatcher = pattern.matcher(getString(config));
+    String dir = partsMatcher.replaceAll("").replace("${topic}", "");
+
+    // check for invalid ${} substitutions
+    Pattern invalid = Pattern.compile("\\$\\{.*}");
+    Matcher invalidMatcher = invalid.matcher(dir);
+    if (invalidMatcher.find()) {
+      throw new ConfigException(
+          String.format(
+              "%s: %s contains an invalid ${} substitution %s",
```
Comment (gharris1727) [inline src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: Can they escape these substitutions like `\$\{topic\}` and have their directory be literally `${topic}/some-topic`? Would this check prevent this use-case? Maybe it's better to be left out.
You can actually generate a list of these substitutions and then shell out to the connect-utils validator. Maybe we can use this if we don't target versions that are included in CP.

File: src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorTestBase.java [hunk 21]
```diff
@@ -73,6 +72,9 @@
     props.put(PartitionerConfig.LOCALE_CONFIG, "en");
     props.put(PartitionerConfig.TIMEZONE_CONFIG, "America/Los_Angeles");
 
+    props.put(HdfsSinkConnectorConfig.TOPIC_REGEX_CAPTURE_GROUP_CONFIG, "(.*)[\\.\\-](.*)");
```
Comment (gharris1727) [inline src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorTestBase.java]: Do these belong in the test base class?
Maybe we can scope these to just the relevant test classes? idk, i'm kinda 50/50 on this.
It does mean that we're not testing with this config disabled, which probably isn't good.

I think you can keep the refactor for topicsDir as a Map, but just disable the config options.

File: src/test/java/io/confluent/connect/hdfs/wal/WALFileTest.java [hunk 22]
```diff
@@ -36,10 +36,12 @@
   @Test
   public void testAppend() throws Exception {
     setUp();
+    properties.put(HdfsSinkConnectorConfig.TOPIC_REGEX_CAPTURE_GROUP_CONFIG, "(.*)");
```
Comment (gharris1727) [inline src/test/java/io/confluent/connect/hdfs/wal/WALFileTest.java]: why is `testAppend` special?

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 23]
```diff
@@ -75,12 +84,30 @@
   private static final String HADOOP_HOME_DOC = "The Hadoop home directory.";
   private static final String HADOOP_HOME_DISPLAY = "Hadoop home directory";
 
+  // Storage group
+  public static final String TOPIC_REGEX_CAPTURE_GROUP_CONFIG = "topic.regex.capture.group";
+  public static final String TOPIC_REGEX_CAPTURE_GROUP_DISPLAY = "Topic Regex Capture Group";
+  public static final String TOPIC_REGEX_CAPTURE_GROUP_DOC = "A regex that matches the entire "
+      + "topic and  specifies what groups to capture in the topic, so when specifying `${1}` in "
+      + "`topics.dir`, `${1}` will refer to the first captured group. Example config value of "
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: I won't be supporting `${0}` but otherwise changed

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 24]
```diff
@@ -75,12 +84,30 @@
...
+      + "`([a-zA-Z]*)_([a-zA-Z]*)` match a topic that is two words delimited by a - and will "
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: making me feel like a certain someone that needs grammarly

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 14 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: I didn't really wanna support `${0}`, `${topic}` is more intuitive I think

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 15 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: ‾\_(ツ)_/‾

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 17 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: yes

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 15 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: `values()` is called on a hashmap so actually order isn't guaranteed

Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: fixed

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 19 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: initially said I wouldn't support but maybe I should

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 20 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: was going to say can't use `connect-utils` but depends on what version - this can be a WIP - this check still catches all `${-n}` values if they are somehow silly enough to put negative numbers. honestly why would they escape and if they do... let it break

File: src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorTestBase.java (hunk 21 above)
Comment (levzem) [reply src/test/java/io/confluent/connect/hdfs/HdfsSinkConnectorTestBase.java]: I put this in the test base class because it is the easiest way to "integration" test this config to make sure it works

File: src/test/java/io/confluent/connect/hdfs/wal/WALFileTest.java (hunk 22 above)
Comment (levzem) [reply src/test/java/io/confluent/connect/hdfs/wal/WALFileTest.java]: it uses a different topic that isn't captured by the normal regex

File: src/main/java/io/confluent/connect/hdfs/DataWriter.java [hunk 25]
```diff
@@ -385,7 +399,7 @@ public void syncWithHive() throws ConnectException {
 
     try {
       for (String topic : topics) {
-        String topicDir = FileUtils.topicDirectory(url, topicsDir, topic);
+        String topicDir = FileUtils.topicDirectory(url, topicDirs.get(topic), topic);
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/DataWriter.java]: this satisfies requirements so I am not sure this is a necessary change. We can easily extend the functionality later if it is required

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 26]
```diff
@@ -458,6 +660,18 @@ public static ConfigDef getConfig() {
...
     );
 
+    visible.define(
+        TOPICS_DIR_CONFIG,
+        Type.STRING,
+        TOPICS_DIR_DEFAULT,
+        Importance.HIGH,
+        TOPICS_DIR_DOC + DIR_REGEX_DOC,
+        "Storage",
+        2,
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: it's the order in `StorageCommonConfig` unfortunately I don't see a better way than hardcoding

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 27]
```diff
@@ -75,12 +84,30 @@
...
+
+  // HDFS Group
   public static final String LOGS_DIR_CONFIG = "logs.dir";
   public static final String LOGS_DIR_DOC =
-      "Top level directory to store the write ahead logs.";
+      "Top level directory to store the write ahead logs." + DIR_REGEX_DOC;
   public static final String LOGS_DIR_DEFAULT = "logs";
   public static final String LOGS_DIR_DISPLAY = "Logs directory";
-
+
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: I don't see a double line in my source code after `LOGS_DIR_DISPLAY`
github shows double lines sometimes incorrectly

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 13 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: this is only run once at ConfigConstruction essentially being static - but I can make it static

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 28]
```diff
@@ -292,9 +341,12 @@ public static ConfigDef newConfigDef() {
   private final StorageCommonConfig commonConfig;
   private final HiveConfig hiveConfig;
   private final PartitionerConfig partitionerConfig;
+  private final Pattern topixRegexCaptureGroup;
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: el oh el

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java (hunk 27 above)
[the remark this answers is not in the document]
Comment (gharris1727) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: It's not a double line, it's a line that was empty before (`\n\n`) and now it has some spaces on it (`\n   \n`)

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 29]
```diff
@@ -399,6 +465,141 @@ public Configuration getHadoopConfiguration() {
...
+      if (!matcher.matches()) {
+        throw new ConfigException(
+            TOPIC_REGEX_CAPTURE_GROUP_CONFIG,
+            topixRegexCaptureGroup.pattern(),
+            String.format("Topic %s does not fully match the specified regex.", topic)
+        );
+      }
+
+      // make sure that all references to captured groups actually exist
+      if (maxIndexParts > 0 && maxIndexParts > matcher.groupCount()) {
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: yes

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 30]
```diff
@@ -399,6 +465,141 @@ public Configuration getHadoopConfiguration() {
...
+
+    return toReplace.isEmpty() ? 0 : Collections.max(toReplace);
+  }
+
+  /**
+   * Validates that the `topics.dir`, `logs.dir`, and `topic.regex.capture.group` configs are all
+   * within valid ranges.
+   */
+  private void validateDirsAndRegex() {
+    if (topicDirGroupsMaxIndex > 0 && topixRegexCaptureGroup == null) {
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: fixed to return -1

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 31]
```diff
@@ -399,6 +465,141 @@ public Configuration getHadoopConfiguration() {
...
+                    + "but actually had %d capture groups.",
+                topic,
+                maxIndexParts,
+                topixRegexCaptureGroup.pattern(),
+                matcher.groupCount()
+            )
+        );
+      }
+
+      for (int index = 0; index < matcher.groupCount() + 1; index++) {
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: I think is an unnecessary micro optimization

File: src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java [hunk 32]
```diff
@@ -399,6 +465,141 @@ public Configuration getHadoopConfiguration() {
...
+  }
+
+  /**
+   * Performs all substitutions and calculates the final directory for a topic
+   *
+   * @param dir - String - the directory to perform substitutions on
+   * @param topic - String - the topic whose directory to find
+   * @return String - the directory name and path
+   */
+  private String getDirFromTopic(String dir, String topic, int maxIndexParts) {
```
[the remark this answers is not in the document]
Comment (levzem) [reply src/main/java/io/confluent/connect/hdfs/HdfsSinkConnectorConfig.java]: it helps with validation instead of recomputing it every time - I don't think we necessarily gain anything by removing the instance variables
